### DISCUSSION TITLE
Mail does not get sent

### DISCUSSION THREAD

**Discussion by SangeloDev** [outline/outline#3308] [Self Hosting]
I've configured my email settings like this in my `.env` file:
```SMTP_HOST=smtp.<domain>
SMTP_PORT=587
SMTP_USERNAME=no-reply@<domain>
SMTP_PASSWORD=<password>
SMTP_FROM_EMAIL=no-reply@<domain>
SMTP_REPLY_EMAIL=no-reply@<domain>
SMTP_TLS_CIPHERS=
SMTP_SECURE=false
```
But I don't get any magic link mails on my test email.
There are no errors in the console.
I've seen other people having issues similar to mine, but the only solution I have found is to use Yahoo, which I don't really like, so I am trying again. Any help appreciated!

**Comment by tommoor** [comment]
Your SMTP server has no SSL configured? What are you using as a provider? Any errors in the Outline server logs?

**Reply by SangeloDev** [reply]
When I enable it, I have some problems with it not connecting or not receiving mails, I am hosting it myself.

**Reply by SangeloDev** [reply]
and also no, the Outline logs are empty

**Reply by almereyda** [reply]
As your mail server uses a non-standard, insecure configuration, this would also be a place to look for error messages related to the sending attempt. Can you find mentions of Outline trying to send mail within your `/var/log/mail.log` or `/var/log/mail.err`?

Having said this, it is strongly recommend that you finish configuring your mail server to a secure setup, with including SPF and DKIM settings for its domain. Insecure transmission of email messages can be considered deprecated nowadays, since this exposes the plain text of every email message to every intermediary on the transmission route.

As per https://nodemailer.com/smtp/, `nodemailer` supports insecure connections when setting the `ignoreTLS` parameter on a transport.

This parameter is not (yet?) implemented in the Outline mailer:

https://github.com/outline/outline/blob/fea5f69a38d19244e70cac2e74e260cb84496a31/server/mailer.tsx#L73-L101

You could monkey-patch this file to include a statement that allows insecure mail connections. The `nodemailer` documentation on the SMTP transport reads:

> - **ignoreTLS** – if this is **true** and **secure** is false then TLS is not used even if the server supports STARTTLS extension
>
> > Setting **secure** to **false** does not mean that you would not use an encrypted connection. Most SMTP servers allow connection upgrade via [STARTTLS](https://tools.ietf.org/html/rfc3207#section-2) command but to use this you have to connect using plaintext first

**Reply by SangeloDev** [reply]
I tried a monkey-patch just now, but still no luck.
```
export class Mailer {
  transporter: any | null | undefined;

  constructor() {
    this.loadTransport();
  }

  async loadTransport() {
    if (process.env.SMTP_HOST) {
      const smtpConfig = {
        host: process.env.SMTP_HOST,
        port: process.env.SMTP_PORT,
        secure:
          "SMTP_SECURE" in process.env
            ? process.env.SMTP_SECURE === "false"
            : process.env.NODE_ENV === "production",
        auth: undefined,
        tls:
          "SMTP_TLS_CIPHERS" in process.env
            ? {
                ciphers: process.env.SMTP_TLS_CIPHERS,
              }
            : undefined,
          ignoreTLS:true
```
I've tried with both ``process.env.SMTP_SECURE === "false"`` and ``process.env.SMTP_SECURE === "true"`` and rebuilt Outline in between changes as well.
I am now completely lost. If it could help somehow, I am using Mailcow inside a docker container as a mail server.

**Comment by almereyda** [comment]
Please note you can increase the loglevel of the Outline application with setting the `DEBUG` variable.

This is what we use:

    DEBUG=cache,presenters,events,emails,mailer,utils,http,server,services

**Comment by Limezy** [comment]
Hi ! For my own instance (running on a Yunohost server) I have made a small patch to ignore TLS. 
TLS was preventing Outline from sending emails on my own installation. I don't need it since the email server is at localhost.
https://github.com/YunoHost-Apps/outline_ynh/blob/v0.62.0-beta2/sources/patches/app-02-email.patch
You could try adding "ignoreTLS:true" at the same place and compile Outline by yourself

_EDIT : this is actually exactly what @almereyda has been proposing to you above_

**Reply by SangeloDev** [reply]
I tried a patch just now, but still no luck.
```
export class Mailer {
  transporter: any | null | undefined;

  constructor() {
    this.loadTransport();
  }

  async loadTransport() {
    if (process.env.SMTP_HOST) {
      const smtpConfig = {
        host: process.env.SMTP_HOST,
        port: process.env.SMTP_PORT,
        secure:
          "SMTP_SECURE" in process.env
            ? process.env.SMTP_SECURE === "false"
            : process.env.NODE_ENV === "production",
        auth: undefined,
        tls:
          "SMTP_TLS_CIPHERS" in process.env
            ? {
                ciphers: process.env.SMTP_TLS_CIPHERS,
              }
            : undefined,
          ignoreTLS:true
```
I've tried with both ``process.env.SMTP_SECURE === "false"`` and ``process.env.SMTP_SECURE === "true"`` and rebuilt Outline in between changes as well.
I am now completely lost. If it could help somehow, I am using Mailcow inside a docker container as a mail server.

**Reply by almereyda** [reply]
You can leave the code about `process.env.SMTP_SECURE` unchanged, but make sure to set `SMTP_SECURE=false` in your environment file. This is also stated in the `nodemailer` documentation. Note the "*and*" and "*false*":

> - **ignoreTLS** – if this is **true** and **secure** is false ...

Maybe this is it?

---

Else, did you manage to triage and isolate the issues that render your Mailcow TLS setup incomplete? If you could create respective issues over there, I'd be happy to follow and help debugging further.

**Reply by dveretennikov1** [reply]
Also office365 smtp works with this settings
SMTP_HOST=smtp.office365.com
SMTP_PORT=587
SMTP_USERNAME=info@xxxxxxx
SMTP_PASSWORD=xxxxxxxxx
SMTP_FROM_EMAIL=info@xxxxxxxxx
SMTP_REPLY_EMAIL=info@xxxxxxxxx
SMTP_TLS_CIPHERS=
SMTP_SECURE=false

I'm tried to use my own smtp server, but got an error because a have only self-signed ssl cert on smtp server

**Reply by SangeloDev** [reply]
> Else, did you manage to triage and isolate the issues that render your Mailcow TLS setup incomplete? If you could create respective issues over there, I'd be happy to follow and help debugging further.

I think my problem's with SSL, I have set up everything as well as I can, so I don't know what else it could be. I'll try to get another certificate and see, but now I don't have much time to do so.

**Reply by almereyda** [reply]
A very dangerous way is to tell Node.js to skip certificate validation altogether exists with [setting `NODE_TLS_REJECT_UNAUTHORIZED=0`](https://nodejs.org/api/cli.html#cli_node_tls_reject_unauthorized_value) in the environment of the Outline container, which is **strongly discouraged**.

Give us a shout when you have time to look at your email setup again.

**Reply by SangeloDev** [reply]
Alright, so, I am hosting a Mailcow Dockerized instance on the same machine.
All of my ports are open, and the server is configured so that it at least functions, here is the most important thing I need to fix according to mail-tester.com:
<img width="1094" alt="image" src="https://user-images.githubusercontent.com/47925793/161219985-122eec6b-c893-49e6-ac9f-1c63d47dc68e.png">

Nonetheless, my setup is working quite fine, I can send and receive e-mails, just not use the outline mailer, or I also had the same problem with another domain and project I am working on. I try to enter the same details I use for the SMTP settings as I use on my Mail client, which I can use flawlessly. I don't really know what I can do here.

**Reply by almereyda** [reply]
DKIM is really good, as is secure transmission of email messages. Please make yourself comfortable with both specifications, and proceed with care to implement them, as they are very useful vor everyone else on the Internet. Imagine someone was MITM your connection to the mail server, got access to valid credentials, and would automatically send Spam messages from your server; or worse.

- DKIM keys will have to be generated on your mail server on a per-domain basis. The public part of the signing key is then entered into the DNS zone for each domain. The web interface of your mail server should give you a hint at how to create these RRs.
- You state your "setup is working *quite* fine" (emphasis by me), which to me reads that you are yourself not perfectly comfortable with it, yet. I'd suggest to try to increase the security of your mail server first, before we attempt to decrease the security of the applications using it.

Your mail client will be unopinionated about the security setting of the connection, as it's primary goal is to get your mail to and from you. Everything else should be handled on the infrastructure level, to reduce the overall attack surface and misuse possibilities of your mail server.

If you give me permission to snoop you, I can try to successfully intercept your mail connections and show how this insecure configuration could be used for malicious behaviour. If you like.

**Reply by SangeloDev** [reply]
I do have everything set up, as you can see in both screenshots below, however I would be happy for you to try out, as I am now curious about how that works :p
![Flameshot-2022-04-02_22-40](https://user-images.githubusercontent.com/47925793/161400577-d3bfdeff-1871-4fcd-8759-5993e91060d6.png)
![Flameshot-2022-04-02_22-42](https://user-images.githubusercontent.com/47925793/161400578-e8299e86-dde9-46e4-87d4-34a5f131917f.png)
I have blurred out the key for obvious reasons, although it is there. I don't know why mail-tester.com says that the DKIM is invalid, maybe because Mailcow self-signs them? Although I don't know how they work that much. I have tried to make it as secure as possible.
Thanks for your help.

**Reply by almereyda** [reply]
It's okay. The domain key selector should be a word, usually it's default, but it can be any that you set. `*` seems weird to me, but if the Mailcow documentation tells you to do so, it might be fine. The Screenshot suggests to use `dkim` as a selector, which would result in a RR for `dkim._domainkey`.

There's also no need to blur out the key, as it is the public key part of a keypair. The private part is used by your server to sign the messages, while the public part is used to verify that signature.

```
$ dig +short TXT dkim._domainkey.sangelo.space
"v=DKIM1;k=rsa;t=s;s=email;p=MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAubZRIR92cZyWj+abDK0zGUKeDipH0bQZxBHlLJ1s4A7YZ4Gxz+Ck1gbZel+Ru2tQR0xgmw52BNXUWF91uoW19qKDrbAi5IFWMv7onEX1CU+8cw9pIQ+bEtxR5W21C8kANytBJsrj61arnNjBJVuA6QhQbYvmyt0zEUqNDBMFWMMcU+5opft7kmu" " Ypm/c9m0S/0BEeROuBAkU4CeZY9V5g+/Uw+On3M34p3UxA8DKxB4TGBxptdUunRW12QtNhJVeEFYmjWTcMioI7K34P4qnQ/JwdS2oNdPK2mCuuapa3Ft3dbvHQRwJwHo23wghYSzSA7SinvMOZznZjT+NPnrFiwIDAQAB"
"v=DKIM1;k=rsa;t=s;s=email;p=MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAubZRIR92cZyWj+abDK0zGUKeDipH0bQZxBHlLJ1s4A7YZ4Gxz+Ck1gbZel+Ru2tQR0xgmw52BNXUWF91uoW19qKDrbAi5IFWMv7onEX1CU+8cw9pIQ+bEtxR5W21C8kANytBJsrj61arnNjBJVuA6QhQbYvmyt0zEUqNDBMFWMMcU+5opft7kmu" "Ypm/c9m0S/0BEeROuBAkU4CeZY9V5g+/Uw+On3M34p3UxA8DKxB4TGBxptdUunRW12QtNhJVeEFYmjWTcMioI7K34P4qnQ/JwdS2oNdPK2mCuuapa3Ft3dbvHQRwJwHo23wghYSzSA7SinvMOZznZjT+NPnrFiwIDAQAB"
```

I think you could revert the `*` subdomain to only answer for `dkim._domainkey` only. It can also be good to review why the records are doubled. But we're getting closer.

After DKIM, there will still be the task to correctly set up secured SMTP and IMAP services, probably with help of Let's Encrypt, in so Outline can talk to your mail server appropriately.

**Reply by SangeloDev** [reply]
> It can also be good to review why the records are doubled. But we're getting closer.

Turns out, the records were really doubled on my Cloudflare panel, so I removed one of the duplicates.

> I think you could revert the * subdomain to only answer for dkim._domainkey only.

Just updated the DNS settings, so it should be changed in a while.

> After DKIM, there will still be the task to correctly set up secured SMTP and IMAP services, probably with help of Let's Encrypt, in so Outline can talk to your mail server appropriately.

I think the problem occurs because I have set up SSL from my NginX panel (I use NginX Proxy Manager to make things faster on other user's ends) and Cloudflare also adds its own, so that might be something?
EDIT: nvm that isn't the problem. Though, my MX record is set up to point to my server's hostname.

**Reply by Limezy** [reply]
@SangeloDev I don't know what's your use case exactly, but seems like a *LOT* of work to get your things done.
FYI I'm using the Yunohost debian-based OS to manage my self-hosted system. 
* Having a top-notch working email config comes by default, only two or three clicks... 
* Having a fully working installation of Outline, including https, MinIO for documents and attachements, Dex for OIDC auth against Yunohost LDAP user base etc.. is also done with only two or three clicks

**Reply by Limezy** [reply]
https://yunohost.org/#
https://github.com/YunoHost-Apps/outline_ynh

**Reply by almereyda** [reply]
Yes, I am also not able to support third-party applications, like Nginx Proxy Manager or Cloudflare, and would suggest you to consult their documentation on how to successfully set up the use case you have in mind.

Somewhere in the Nginx Proxy Manager could be something that generates Let's Encrypt certificates. These can be reused by your MTA and MDA, which also need to be reloaded through a hook fired by certificate renewal. The documentation of your software vendor will (have to) include hints about how to integrate with the chosen Let's Encrypt/ACME client, and how to reload the MX services upon renewal.
